tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -1601,180 +1655,173 @@ uint32_t SPIRVEmitter::doCXXMemberCallExpr(const CXXMemberCallExpr *expr) {
-    case IntrinsicOp::MOP_Sample:
-    case IntrinsicOp::MOP_Gather: {
-      // Signatures:
-      // DXGI_FORMAT Object.Sample(sampler_state S,
-      //                           float Location
-      //                           [, int Offset]);
-      //
-      // <Template Type>4 Object.Gather(sampler_state S,
-      //                                float2|3|4 Location
-      //                                [, int2 Offset]);
-
-      const auto *imageExpr = expr->getImplicitObjectArgument();
-
-      const uint32_t imageType =
-          typeTranslator.translateType(imageExpr->getType());
-
-      const uint32_t image = loadIfGLValue(imageExpr);
-      const uint32_t sampler = doExpr(expr->getArg(0));
-      const uint32_t coordinate = doExpr(expr->getArg(1));
-      // .Sample()/.Gather() has a third optional paramter for offset.
-      uint32_t constOffset = 0, varOffset = 0;
-      handleOffset(expr, 2, &constOffset, &varOffset);
+  const FunctionDecl *callee = expr->getDirectCallee();
+  const auto retType = typeTranslator.translateType(callee->getReturnType());
 
-      if (opcode == static_cast<uint32_t>(IntrinsicOp::MOP_Sample)) {
-        return theBuilder.createImageSample(
-            retType, imageType, image, sampler, coordinate, /*bias*/ 0,
-            /*lod*/ 0, std::make_pair(0, 0), constOffset, varOffset);
-      } else {
-        return theBuilder.createImageGather(
-            retType, imageType, image, sampler, coordinate,
-            // .Gather() doc says we return four components of red data.
-            theBuilder.getConstantInt32(0), constOffset, varOffset);
-      }
-    }
-    case IntrinsicOp::MOP_SampleBias:
-    case IntrinsicOp::MOP_SampleLevel: {
-      // Signatures:
-      // DXGI_FORMAT Object.SampleBias(sampler_state S,
-      //                               float Location,
-      //                               float Bias
-      //                               [, int Offset]);
-      //
-      // DXGI_FORMAT Object.SampleLevel(sampler_state S,
-      //                                float Location,
-      //                                float LOD
-      //                                [, int Offset]);
-      const auto *imageExpr = expr->getImplicitObjectArgument();
-
-      const uint32_t imageType =
-          typeTranslator.translateType(imageExpr->getType());
-
-      const uint32_t image = loadIfGLValue(imageExpr);
-      const uint32_t sampler = doExpr(expr->getArg(0));
-      const uint32_t coordinate = doExpr(expr->getArg(1));
-      uint32_t lod = 0;
-      uint32_t bias = 0;
-      if (opcode == static_cast<uint32_t>(IntrinsicOp::MOP_SampleBias)) {
-        bias = doExpr(expr->getArg(2));
-      } else {
-        lod = doExpr(expr->getArg(2));
-      }
-      // .Bias()/.SampleLevel() has a fourth optional paramter for offset.
-      uint32_t constOffset = 0, varOffset = 0;
-      handleOffset(expr, 3, &constOffset, &varOffset);
+  switch (opcode) {
```

Comment:
FYI: There are no changes to the `switch` statement except that the `default` case is moved out.